File: frontend/src/api/notifications.ts
import client from './client'

export interface Channel {
  id: string
  kind: string
  target: string | null
  is_enabled: boolean
}

export interface AlertRule {
  id: string
  type: string
  car_id: number | null
  is_enabled: boolean
  params: Record<string, unknown>
  last_fired_at: string | null
}

export interface SmtpOut {
  smtp_host: string | null
  smtp_port: number | null
  smtp_username: string | null
  smtp_from: string | null
  smtp_use_tls: boolean
  password_set: boolean
}

export async function listChannels(): Promise<Channel[]> {
  const res = await client.get<Channel[]>('/notifications/channels')
  return res.data
}

export async function updateChannel(id: string, data: { target?: string | null; is_enabled?: boolean }): Promise<Channel> {
  const res = await client.patch<Channel>(`/notifications/channels/${id}`, data)
  return res.data
}

export async function sendTest(): Promise<void> {
  await client.post('/notifications/test')
}

export async function listAlerts(): Promise<AlertRule[]> {
  const res = await client.get<AlertRule[]>('/notifications/alerts')
  return res.data
}

export async function createAlert(data: {
  type: string
  car_id?: number | null
  is_enabled?: boolean
  params?: Record<string, unknown>
}): Promise<AlertRule> {
  const res = await client.post<AlertRule>('/notifications/alerts', data)
  return res.data
}

export async function updateAlert(
  id: string,
  data: { is_enabled?: boolean; car_id?: number | null; params?: Record<string, unknown> },
): Promise<AlertRule> {
  const res = await client.patch<AlertRule>(`/notifications/alerts/${id}`, data)
  return res.data
}

export async function deleteAlert(id: string): Promise<void> {
  await client.delete(`/notifications/alerts/${id}`)
}

export async function getSmtp(): Promise<SmtpOut> {
  const res = await client.get<SmtpOut>('/notifications/smtp')
  return res.data
}

export async function updateSmtp(data: {
  smtp_host?: string | null
  smtp_port?: number | null
  smtp_username?: string | null
  smtp_password?: string | null
  smtp_from?: string | null
  smtp_use_tls?: boolean
}): Promise<SmtpOut> {
  const res = await client.put<SmtpOut>('/notifications/smtp', data)
  return res.data
}

File: backend/schemas/notifications.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field


class ChannelOut(BaseModel):
    id: uuid.UUID
    kind: str
    target: str | None = None
    is_enabled: bool

    model_config = {"from_attributes": True}


class ChannelUpdate(BaseModel):
    target: str | None = None
    is_enabled: bool | None = None


class AlertRuleIn(BaseModel):
    type: str
    car_id: int | None = None
    is_enabled: bool = True
    params: dict = Field(default_factory=dict)


class AlertRuleUpdate(BaseModel):
    is_enabled: bool | None = None
    car_id: int | None = None
    params: dict | None = None


class AlertRuleOut(BaseModel):
    id: uuid.UUID
    type: str
    car_id: int | None = None
    is_enabled: bool
    params: dict
    last_fired_at: datetime | None = None

    model_config = {"from_attributes": True}


class SmtpConfig(BaseModel):
    smtp_host: str | None = None
    smtp_port: int | None = None
    smtp_username: str | None = None
    smtp_password: str | None = None   # write-only; never returned
    smtp_from: str | None = None
    smtp_use_tls: bool = True


class SmtpOut(BaseModel):
    smtp_host: str | None = None
    smtp_port: int | None = None
    smtp_username: str | None = None
    smtp_from: str | None = None
    smtp_use_tls: bool = True
    password_set: bool = False
